### PR TITLE
Add option for mutex timeout in distributed optimizer backward hook

### PR DESCRIPTION
**Pull request by minitu**
# What does this PR do ?

Modified version of https://github.com/NVIDIA/NeMo/pull/9084, to debug a hang at https://github.com/NVIDIA/NeMo/blob/f658b6f0445403c338c7371941b1fe644832df48/nemo/core/optim/distributed_adam.py#L131

**Collection**: NLP

# Changelog 
- Add specific line by line info of high level changes in this PR.

# Usage
Run GPT, e.g. with the config at https://github.com/NVIDIA/NeMo/blob/main/examples/nlp/language_modeling/conf/megatron_gpt_config.yaml.

Enable the distributed optimizer with `model.optim.name=distributed_fused_adam` and set the timeout with `model.optim.lock_timeout=<seconds>`.

# Jenkins CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

There's no need to comment `jenkins` on the PR to trigger Jenkins CI.
The GitHub Actions CI will run automatically when the PR is opened.
To run CI on an untrusted fork, a NeMo user with write access must click "Approve and run".

# Before your PR is "Ready for review"
**Pre checks**:
- [x] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [x] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [x] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

**Comment by minitu** [comment]
@ShriyaPalsamudram Tested on Eos single-node, the updated code spits out a RuntimeError when the lock is not obtained within the given timeout period.

**Review by ShriyaPalsamudram — DISMISSED** [review]

**Comment by layalir** [comment]
How were you able to create a timeout situation in one node? Did you set the timer to very very small value?

File: nemo/core/optim/distributed_adam.py [hunk 1]
```diff
@@ -152,6 +156,25 @@ def __init__(
...
+                result = self._lock.acquire(timeout=lock_timeout)
+                try:
+                    yield result
+                finally:
+                    if result:
+                        # Acquired lock before timeout
+                        self._lock.release()
+                    else:
+                        # Failed to acquire lock before timeout
+                        print(f'MegatronDistributedFusedAdam: Failed to acquire lock within {lock_timeout} seconds.')
```
Comment (ericharper) [inline nemo/core/optim/distributed_adam.py:174]: this should use `logging`

Comment (minitu) [reply nemo/core/optim/distributed_adam.py:174]: The issue is that we want the print to happen even with NeMo logging disabled, for hang debugging purposes. @ShriyaPalsamudram

**Pull request merged**